File: applications/rtkbackprojections/rtkbackprojections.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkbackprojections_ggo.h"
#include "rtkGgoFunctions.h"

#include "rtkThreeDCircularProjectionGeometryXMLFile.h"
#include "rtkFDKBackProjectionImageFilter.h"
#include "rtkFDKWarpBackProjectionImageFilter.h"
#include "rtkJosephBackProjectionImageFilter.h"
#include "rtkJosephBackAttenuatedProjectionImageFilter.h"
#include "rtkZengBackProjectionImageFilter.h"
#ifdef RTK_USE_CUDA
#  include "rtkCudaFDKBackProjectionImageFilter.h"
#  include "rtkCudaBackProjectionImageFilter.h"
#  include "rtkCudaRayCastBackProjectionImageFilter.h"
#endif
#include "rtkCyclicDeformationImageFilter.h"

#include <itkImageFileReader.h>
#include <itkImageFileWriter.h>

int
main(int argc, char * argv[])
{
  GGO(rtkbackprojections, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3;
#ifdef RTK_USE_CUDA
  using OutputImageType = itk::CudaImage<OutputPixelType, Dimension>;
#else
  using OutputImageType = itk::Image<OutputPixelType, Dimension>;
#endif
  // Geometry
  if (args_info.verbose_flag)
    std::cout << "Reading geometry information from " << args_info.geometry_arg << "..." << std::flush;
  rtk::ThreeDCircularProjectionGeometry::Pointer geometry;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(geometry = rtk::ReadGeometry(args_info.geometry_arg));
  if (args_info.verbose_flag)
    std::cout << " done." << std::endl;

  // Create an empty volume
  using ConstantImageSourceType = rtk::ConstantImageSource<OutputImageType>;
  auto constantImageSource = ConstantImageSourceType::New();
  rtk::SetConstantImageSourceFromGgo<ConstantImageSourceType, args_info_rtkbackprojections>(constantImageSource,
                                                                                            args_info);

  // Projections reader
  using ReaderType = rtk::ProjectionsReader<OutputImageType>;
  auto reader = ReaderType::New();
  rtk::SetProjectionsReaderFromGgo<ReaderType, args_info_rtkbackprojections>(reader, args_info);
  TRY_AND_EXIT_ON_ITK_EXCEPTION(reader->Update())

  OutputImageType::Pointer attenuationMap;
  if (args_info.attenuationmap_given)
  {
    if (args_info.verbose_flag)
      std::cout << "Reading attenuation map " << args_info.attenuationmap_arg << "..." << std::endl;
    // Read an existing image to initialize the attenuation map
    attenuationMap = itk::ReadImage<OutputImageType>(args_info.attenuationmap_arg);
  }

  // Create back projection image filter
  if (args_info.verbose_flag)
    std::cout << "Backprojecting volume..." << std::endl;
  rtk::BackProjectionImageFilter<OutputImageType, OutputImageType>::Pointer bp;

  // In case warp backprojection is used, we create a deformation
  using DVFPixelType = itk::Vector<float, 3>;
  using DeformationType = rtk::CyclicDeformationImageFilter<itk::Image<DVFPixelType, 4>, itk::Image<DVFPixelType, 3>>;
  auto def = DeformationType::New();

  switch (args_info.bp_arg)
  {
    case (bp_arg_VoxelBasedBackProjection):
      bp = rtk::BackProjectionImageFilter<OutputImageType, OutputImageType>::New();
      break;
    case (bp_arg_FDKBackProjection):
      bp = rtk::FDKBackProjectionImageFilter<OutputImageType, OutputImageType>::New();
      break;
    case (bp_arg_FDKWarpBackProjection):
      if (!args_info.signal_given || !args_info.dvf_given)
      {
        std::cerr << "FDKWarpBackProjection requires input 4D deformation "
                  << "vector field and signal file names" << std::endl;
        return EXIT_FAILURE;
      }
      def->SetInput(itk::ReadImage<DeformationType::InputImageType>(args_info.dvf_arg));
      bp = rtk::FDKWarpBackProjectionImageFilter<OutputImageType, OutputImageType, DeformationType>::New();
      def->SetSignalFilename(args_info.signal_arg);
      dynamic_cast<rtk::FDKWarpBackProjectionImageFilter<OutputImageType, OutputImageType, DeformationType> *>(
        bp.GetPointer())
        ->SetDeformation(def);
      break;
    case (bp_arg_Joseph):
      bp = rtk::JosephBackProjectionImageFilter<OutputImageType, OutputImageType>::New();
      break;
    case (bp_arg_JosephAttenuated):
      bp = rtk::JosephBackAttenuatedProjectionImageFilter<OutputImageType, OutputImageType>::New();
      break;
    case (bp_arg_Zeng):
      bp = rtk::ZengBackProjectionImageFilter<OutputImageType, OutputImageType>::New();
      break;
    case (bp_arg_CudaFDKBackProjection):
#ifdef RTK_USE_CUDA
      bp = rtk::CudaFDKBackProjectionImageFilter::New();
#else
      std::cerr << "The program has not been compiled with cuda option" << std::endl;
      return EXIT_FAILURE;
#endif
      break;
    case (bp_arg_CudaBackProjection):
#ifdef RTK_USE_CUDA
      bp = rtk::CudaBackProjectionImageFilter<OutputImageType>::New();
#else
      std::cerr << "The program has not been compiled with cuda option" << std::endl;
      return EXIT_FAILURE;
#endif
      break;
    case (bp_arg_CudaRayCast):
#ifdef RTK_USE_CUDA
      bp = rtk::CudaRayCastBackProjectionImageFilter::New();
#else
      std::cerr << "The program has not been compiled with cuda option" << std::endl;
      return EXIT_FAILURE;
#endif
      break;
    default:
      std::cerr << "Unhandled --method value." << std::endl;
      return EXIT_FAILURE;
  }

  bp->SetInput(constantImageSource->GetOutput());
  bp->SetInput(1, reader->GetOutput());
  if (args_info.attenuationmap_given)
    bp->SetInput(2, attenuationMap);
  if (args_info.sigmazero_given && args_info.bp_arg == bp_arg_Zeng)
    dynamic_cast<rtk::ZengBackProjectionImageFilter<OutputImageType, OutputImageType> *>(bp.GetPointer())
      ->SetSigmaZero(args_info.sigmazero_arg);
  if (args_info.alphapsf_given && args_info.bp_arg == bp_arg_Zeng)
    dynamic_cast<rtk::ZengBackProjectionImageFilter<OutputImageType, OutputImageType> *>(bp.GetPointer())
      ->SetAlpha(args_info.alphapsf_arg);
  bp->SetGeometry(geometry);
  TRY_AND_EXIT_ON_ITK_EXCEPTION(bp->Update())

  // Write
  if (args_info.verbose_flag)
    std::cout << "Writing... " << std::endl;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(bp->GetOutput(), args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtkbackprojections/rtkbackprojections.py
#!/usr/bin/env python
import sys
import argparse
import itk
from itk import RTK as rtk


def build_parser():
    parser = argparse.ArgumentParser(
        description="Backprojects a volume according to a geometry file.",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )

    # General arguments
    parser.add_argument(
        "--verbose", "-v", help="Verbose execution", action="store_true"
    )
    parser.add_argument(
        "--geometry", "-g", help="XML geometry file name", type=str, required=True
    )
    parser.add_argument(
        "--output", "-o", help="Output volume file name", type=str, required=True
    )

    # Projectors group
    rtkprojectors_group = parser.add_argument_group("Projectors")
    rtkprojectors_group.add_argument(
        "--bp",
        help="Backprojection method",
        type=str,
        choices=[
            "VoxelBasedBackProjection",
            "FDKBackProjection",
            "FDKWarpBackProjection",
            "Joseph",
            "JosephAttenuated",
            "Zeng",
            "CudaFDKBackProjection",
            "CudaBackProjection",
            "CudaRayCast",
        ],
        default="VoxelBasedBackProjection",
    )

    rtkprojectors_group.add_argument(
        "--attenuationmap", help="Attenuation map for attenuation correction", type=str
    )
    rtkprojectors_group.add_argument(
        "--sigmazero",
        help="PSF value at a distance of 0 meter of the detector",
        type=float,
    )
    rtkprojectors_group.add_argument(
        "--alphapsf", help="Slope of the PSF against the detector distance", type=float
    )

    # Warped backprojections group
    Warped_backprojection_group = parser.add_argument_group(
        "Warped backprojections group"
    )
    Warped_backprojection_group.add_argument(
        "--signal", help="Signal file name", type=str
    )
    Warped_backprojection_group.add_argument("--dvf", help="Input 4D DVF", type=str)

    rtk.add_rtk3Doutputimage_group(parser)
    rtk.add_rtkinputprojections_group(parser)

    # Parse the command line arguments
    return parser


def process(args_info: argparse.Namespace):
    # Define output pixel type and dimension
    OutputPixelType = itk.F
    Dimension = 3
    OutputImageType = itk.Image[OutputPixelType, Dimension]
    if hasattr(itk, "CudaImage"):
        OutputCudaImageType = itk.CudaImage[OutputPixelType, Dimension]

    # Geometry
    if args_info.verbose:
        print(f"Reading geometry from {args_info.geometry}...")
    geometry = rtk.read_geometry(args_info.geometry)
    if args_info.verbose:
        print(f"done.")

    # Create empty volume
    constantImageSource = rtk.ConstantImageSource[OutputImageType].New()
    rtk.SetConstantImageSourceFromArgParse(constantImageSource, args_info)

    # Projections reader
    reader = rtk.ProjectionsReader[itk.Image[itk.F, 3]].New()
    rtk.SetProjectionsReaderFromArgParse(reader, args_info)
    reader.Update()

    attenuation_map = None
    if args_info.attenuationmap:
        if args_info.verbose:
            print(f"Reading attenuation map from {args_info.attenuationmap}...")
        # Read an existing image to initialize the attenuation map
        attenuation_map = itk.imread(args_info.attenuationmap)

    # In case warp backprojection is used, we create a deformation
    DVFPixelType = itk.Vector[itk.F, 3]
    DVFImageSequenceType = itk.Image[DVFPixelType, 4]
    DVFImageType = itk.Image[DVFPixelType, 3]
    DeformationType = rtk.CyclicDeformationImageFilter[
        DVFImageSequenceType, DVFImageType
    ]

    bp = None
    if args_info.bp == "VoxelBasedBackProjection":
        bp = rtk.BackProjectionImageFilter[OutputImageType, OutputImageType].New()
    elif args_info.bp == "FDKBackProjection":
        bp = rtk.FDKBackProjectionImageFilter[OutputImageType, OutputImageType].New()
    elif args_info.bp == "FDKWarpBackProjection":
        if not args_info.dvf or not args_info.signal:
            raise ValueError(
                "FDKWarpBackProjection requires input 4D deformation vector field and signal file names"
            )
        deformation = rtk.CyclicDeformationImageFilter[
            DVFImageSequenceType, DVFImageType
        ].New()
        deformation.SetInput(itk.imread(args_info.dvf))
        bp = rtk.FDKWarpBackProjectionImageFilter[
            OutputImageType, OutputImageType, DeformationType
        ].New()
        deformation.SetSignalFilename(args_info.signal)
        bp.SetDeformation(deformation)
    elif args_info.bp == "Joseph":
        bp = rtk.JosephBackProjectionImageFilter[OutputImageType, OutputImageType].New()
    elif args_info.bp == "JosephAttenuated":
        bp = rtk.JosephBackAttenuatedProjectionImageFilter[
            OutputImageType, OutputImageType
        ].New()
    elif args_info.bp == "Zeng":
        bp = rtk.ZengBackProjectionImageFilter[OutputImageType, OutputImageType].New()
        if args_info.sigmazero:
            bp.SetSigmaZero(args_info.sigmazero)
        if args_info.alphapsf:
            bp.SetAlpha(args_info.alphapsf)

    elif args_info.bp == "CudaFDKBackProjection":
        if hasattr(itk, "CudaImage"):
            bp = rtk.CudaFDKBackProjectionImageFilter.New()
        else:
            print("The program has not been compiled with cuda option")
            sys.exit(1)
    elif args_info.bp == "CudaBackProjection":
        if hasattr(itk, "CudaImage"):
            bp = rtk.CudaBackProjectionImageFilter[OutputCudaImageType].New()
        else:
            print("The program has not been compiled with cuda option")
            sys.exit(1)
    elif args_info.bp == "CudaRayCast":
        if hasattr(itk, "CudaImage"):
            bp = rtk.CudaRayCastBackProjectionImageFilter.New()
        else:
            print("The program has not been compiled with cuda option")
            sys.exit(1)

    if args_info.bp in ["CudaFDKBackProjection", "CudaBackProjection", "CudaRayCast"]:
        bp.SetInput(itk.cuda_image_from_image(constantImageSource.GetOutput()))
        bp.SetInput(1, itk.cuda_image_from_image(reader.GetOutput()))
        if attenuation_map:
            bp.SetInput(2, itk.cuda_image_from_image(attenuation_map))
    else:
        bp.SetInput(constantImageSource.GetOutput())
        bp.SetInput(1, reader.GetOutput())
        if attenuation_map:
            bp.SetInput(2, attenuation_map)

    bp.SetGeometry(geometry)

    bp.Update()

    # Write
    if args_info.verbose:
        print(f"Writing output to {args_info.output}...")

    writter = itk.ImageFileWriter[OutputImageType].New()
    writter.SetFileName(args_info.output)
    writter.SetInput(bp.GetOutput())
    writter.Update()

    if args_info.verbose:
        print("Processing completed successfully.")


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
